### ISSUE TITLE
Storybook broken from >=6.0.8 because React.ComponentProps cannot derive all props

### ISSUE THREAD

**Issue by alex2** [styled-components/styled-components#4225]
## System:
 - OS: macOS 14.1.1
 - CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
 - Memory: 3.95 GB / 64.00 GB
 - Shell: 5.9 - /bin/zsh
## Binaries:
 - Node: 18.16.0 - ~/.nvm/versions/node/v18.16.0/bin/node
 - Yarn: 1.22.19 - ~/.yarn/bin/yarn
 - npm: 9.5.1 - ~/.nvm/versions/node/v18.16.0/bin/npm
 - pnpm: 8.6.6 - ~/.nvm/versions/node/v18.16.0/bin/pnpm
 - bun: 1.0.15 - ~/.bun/bin/bun
## npmPackages:
 - styled-components: latest => 6.1.1 

## Reproduction

```tsx
import styled from "styled-components";

const A = styled.div<{ foo?: string }>``;

const B = styled(A)<{ bar?: string }>``;

// typescript has no issue with the "bar" prop and it is typed correctly
const C = () => <B foo="one" bar="two" />

// Typescript error: "Property 'bar' does not exist on type ..."
// This works as expected on v6.0.7 but is broken in any version since
type T = React.ComponentProps<typeof B>["bar"]
```

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcARFDvmQNwBQokscqMAngDZYAmcRJ5Vpx4BaXCUgA7LJJipadOuMms4AQTgBeFuy7cAdN2AA3ADwBvPhAgB+AFw6owSQHM4AXwB8AA2-0lECrwAEJaOsLcABRqAJQWcABGyFD2js5uXr7+yqoAwmGRMVqecKahBNaaZIFYZInJVTAA7hB1APSeim1tcAAiEFjokhDwqAAWEE0ocOxgg7hOYPBYUMRQcM5wxgBs+gAM+gDsdLNYcAAqYdh4MPq5EjWyAArEYKimpxAEcMGeANpkJJQMgAXSAA

## Steps to reproduce

Using styled-components v6.0.8 or above, up to `@ latest` results in props from wrapped types (including those introduced by `attrs` not being picked up by `React.ComponentProps`.

This results in a number of issues with tooling - most significantly for us, storybook which derives the props using `StoryMeta` to type its story args, meaning all stories now show type errors.

## Expected Behavior

Component B wrapping component A and introducing prop "bar" should type correctly and function with `React.ComponentProps` type.

## Actual Behavior

Typescript shows an error when props are derived this way.

## Extra Information

This seems to be vaguely related to #4165 and #4219 but neither highlight the React.ComponentProps case.

**Comment by joshuaellis** [comment]
You can also see the error here – https://codesandbox.io/p/sandbox/cold-sun-8rn4s7?file=%2Fsrc%2FApp.tsx%3A10%2C11 where the `Box3` component has not inherited all the props it should have.

**Comment by alex2** [comment]
This issue is also raised in a comment here https://github.com/styled-components/styled-components/issues/4174#issuecomment-1731339423

**Comment by justinbhopper** [comment]
The issue specifically began here when support for static properties was added #4141 in order to fix issue #4104

**Comment by quantizor** [comment]
> This issue is also raised in a comment here [#4174 (comment)](https://github.com/styled-components/styled-components/issues/4174#issuecomment-1731339423)

This scenario is fixed by 6.1.3 (thanks @justinbhopper!)

**Comment by quantizor** [comment]
I think the OP issue might be resolved as well by 6.1.3 @alex2, could you verify and close if so? Thanks!
